packages/adapters-library/src/adapters/aave-v3/products/rewards/aaveV3RewardsAdapter.ts

```diff
@@ -0,0 +1,400 @@
+  }
+
+  @CacheToDb
+  async getProtocolTokens(): Promise<ProtocolToken[]> {
+    const rewardTokenAddresses = await this.incentivesContract.getRewardsList()
+
+    const rewardTokens = await Promise.all(
+      rewardTokenAddresses.map((rewardToken) => {
+        return this.helpers.getTokenMetadata(rewardToken)
+      }),
+    )
+
+    return [
+      { ...this.INCENTIVES_CONTRACT_DETAILS, underlyingTokens: rewardTokens },
+    ]
+  }
+
+  private detectPositionEventSignature(
+    eventSignature = 'Transfer(address,address,uint256)',
+  ): string {
+    return ethers.id(eventSignature)
+  }
+
+  /**
+   * Checks if the user has ever opened a position in AaveV3
+   * Return AToken addresses if found
+   */
+  private async openPositions(userAddress: string): Promise<string[]> {
+    const topic0 = this.detectPositionEventSignature()
+    const topic1 = ethers.zeroPadValue(ZERO_ADDRESS, 32)
+    const topic2 = ethers.zeroPadValue(userAddress, 32)
+
+    const logs = await this.provider.getLogs({
+      topics: [topic0, topic1, topic2],
+      fromBlock: '0x',
+      toBlock: 'latest',
+    })
+
+    const aTokenAddresses = await this.getAaveTokenAddresses()
+
+    const filteredLogs = logs
+      .filter((log) => aTokenAddresses.includes(log.address))
+      .map((log) => log.address)
+
+    return [...new Set(filteredLogs)]
+  }
+
+  async getPositions(input: GetPositionsInput): Promise<ProtocolPosition[]> {
+    // Fetch reward contract details
+    const protocolTokens = await this.getProtocolTokens()
+    const protocolToken = protocolTokens[0]
+
+    const { userAddress, blockNumber, protocolTokenAddresses } = input
+
+    // If no protocolToken is found throw an error
+    if (!protocolToken) {
+      throw new Error('No protocol token found')
+    }
+
+    // If protocolTokenAddresses are provided, check if this protocolToken is in the list
+    if (
+      protocolTokenAddresses &&
+      protocolTokenAddresses.length > 0 &&
+      !protocolTokenAddresses.includes(protocolToken.address)
+    ) {
+      return []
+    }
+
+    // Check if user has ever opened a position in AaveV3
+    const addressFilter = await this.openPositions(userAddress)
+
+    // Return empty array if user has never opened a AaveV3 position
+    if (!addressFilter?.length) {
+      return []
+    }
+
+    // Fetch and filter underlying tokens with non-zero rewards in parallel
+    const underlyingTokens = await this.getClaimableUnderlyingTokens(
+      userAddress,
+      protocolToken.underlyingTokens,
+      blockNumber,
+    )
+
+    // If no claimable underlying tokens are found, return empty
+    if (underlyingTokens.length === 0) {
+      return []
+    }
+
+    // Return the final ProtocolPosition object
+    return [
+      {
+        type: TokenType.Protocol,
+        balanceRaw: 1n, // choose 1 here as a zero value may cause the position to be ignored on UIs our adapters currently expecting a protocol token but on contract positions there is no token
```

Comment:
Oof

Reply:
I think this is a bad idea too.

If the client is ignoring it because it shows 0, it's the client's problem.

Reply:
We have an open ticket to deal with this sort of scenarios, maybe the answer is for contract positions to not have a balance at all, just underlyings.

Reply:
Hey, 

Did you guys know the balance means different things on a standard erc20 token vs a standard erc721

Reminder:
```
ERC-20: Represents a countable, divisible balance of identical tokens.
ERC-721: Represents ownership of distinct, indivisible assets where each token is unique and tracked by its own ID.
```

@bergarces maybe it does make sense to have a balance but for contract and nft positions it represents the number of distinct positions held.

Reply:
Thoughts?

So 1n should be underlying.length and decimals 1

Reply:
maybe for contract and nft positions we should drop the decimals but yeah not sure atm